operator/metadata/updater.go

```diff
@@ -0,0 +1,260 @@
+}
+
+type shareStorage interface {
+	List(txn basedb.Reader, filters ...registrystorage.SharesFilter) []*ssvtypes.SSVShare
+	Range(txn basedb.Reader, fn func(*ssvtypes.SSVShare) bool)
+	UpdateValidatorsMetadata(map[spectypes.ValidatorPK]*beacon.ValidatorMetadata) error
+}
+
+type fetcher interface {
+	Fetch(ctx context.Context, pubKeys []spectypes.ValidatorPK) (Validators, error)
+}
+
+func NewUpdater(
+	logger *zap.Logger,
+	shareStorage shareStorage,
+	beaconNetwork beacon.BeaconNetwork,
+	beaconNode beacon.BeaconNode,
+	opts ...Option,
+) *Updater {
+	u := &Updater{
+		logger:            logger,
+		shareStorage:      shareStorage,
+		beaconNetwork:     beaconNetwork,
+		fetcher:           NewFetcher(logger, beaconNode),
+		updateInterval:    defaultUpdateInterval,
+		streamInterval:    defaultStreamInterval,
+		updateSendTimeout: defaultUpdateSendTimeout,
+	}
+
+	for _, opt := range opts {
+		opt(u)
+	}
+
+	return u
+}
+
+type Option func(*Updater)
+
+func WithUpdateInterval(interval time.Duration) Option {
+	return func(u *Updater) {
+		u.updateInterval = interval
+	}
+}
+
+func (u *Updater) RetrieveInitialMetadata(ctx context.Context) (map[spectypes.ValidatorPK]*beacon.ValidatorMetadata, error) {
+	// Load non-liquidated shares.
+	shares := u.shareStorage.List(nil, registrystorage.ByNotLiquidated())
+	if len(shares) == 0 {
+		u.logger.Info("could not find validators")
+		return nil, nil
+	}
+
+	needToUpdate := false
+	allPubKeys := make([]spectypes.ValidatorPK, 0, len(shares))
+	for _, share := range shares {
+		allPubKeys = append(allPubKeys, share.ValidatorPubKey)
+		if !share.HasBeaconMetadata() {
+			needToUpdate = true
+		}
+	}
+
+	if !needToUpdate {
+		return nil, nil
+	}
+
+	return u.Update(ctx, allPubKeys)
+}
+
+func (u *Updater) Update(ctx context.Context, pubKeys []spectypes.ValidatorPK) (Validators, error) {
+	fetchStart := time.Now()
+	metadata, err := u.fetcher.Fetch(ctx, pubKeys)
+	if err != nil {
+		u.logger.Error("failed to fetch initial validators metadata",
+			zap.Int("shares", len(pubKeys)),
+			fields.Took(time.Since(fetchStart)),
+			zap.Error(err),
+		)
+		return nil, fmt.Errorf("fetch metadata: %w", err)
+	}
+
+	u.logger.Debug("🆕 fetched metadata",
+		fields.Took(time.Since(fetchStart)),
+		zap.Int("count", len(metadata)),
+		zap.Int("shares", len(pubKeys)),
+	)
+
+	updateStart := time.Now()
+	// TODO: Refactor share storage to support passing context.
+	if err := u.shareStorage.UpdateValidatorsMetadata(metadata); err != nil {
+		u.logger.Error("failed to update validators metadata after setup",
+			zap.Int("shares", len(pubKeys)),
+			fields.Took(time.Since(updateStart)),
+			zap.Error(err),
+		)
+		return metadata, fmt.Errorf("update metadata: %w", err)
+	}
+
+	u.logger.Debug("🆕 updated validators metadata in storage",
+		fields.Took(time.Since(updateStart)),
+		zap.Int("count", len(metadata)),
+		zap.Int("shares", len(pubKeys)),
+	)
+	return metadata, nil
+}
+
+type Update struct {
```

Comment:
Initially, I was thinking about naming it `MetadataUpdate`, but then it starts with the package name. And the package route used to be `validator/metadata`, but then I moved it. I'll think about this once again, maybe I'll move this or rename to something like `ValidatorUpdate`

Reply:
This pretty much ties with the other suggestion from below: https://github.com/ssvlabs/ssv/pull/1805#discussion_r1827875337

I'm not sure what the best combination of **package name + struct(s) name(s)** would be (ideally, it would accommodate future developments too - if we want to add a new syncer other than ValidatorMetadataUpdater) - it just didn't feel like the one I saw during that last review was the optimal one